File: nvpl_blas/f77/strsv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     STRSV 
!
!******************************************************************************
      program   STRSV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, i, lda
      parameter       (n=2, incx=1, lda=2)
      integer         xsize, apsize
      parameter       (xsize=2+(n-1)*abs(incx))
      real            x(xsize), a(lda,n)

*       External Subroutines
      external        print_svector, fill_svector,
     &                print_smatrix, fill_smatrix
      character*1     uplo, trans, diag
      parameter       (uplo='U', trans='N', diag='N')
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, lda, uplo, trans, diag, incx
      call fill_smatrix(a,n,n,lda,uplo,diag)
      call fill_svector(x,n,incx)
      call print_svector(x,n,incx,'X')
      call print_smatrix(a,n,n,lda,'A')
*
*      Call STRSV subroutine
      call STRSV(uplo,trans,diag,n,a,lda,x,incx)
*
      print*
      print 102
      call print_svector(x,n,incx,'X')

      stop
 99   format('Example: STRSV for solving a system of linear equations ' 
     &       'whose coefficients are in a triangular matrix')
 100  format('#### args: n=',i1,', lda=',i1,', uplo=',a1,', trans=',a1,
     &       ', diag=',a1,', incx=',i1)
 102  format('The result of STRSV: ')
      end

File: nvpl_blas/c/strsv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_strsv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t incX = 1;
    nvpl_int_t lda = N;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_TRANSPOSE trans = CblasNoTrans;
    enum CBLAS_DIAG diag = CblasNonUnit;
    float * A = NULL;
    float * X = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);

    printf("\nExample: cblas_strsv for solving a system of linear equations whose coefficients are in a triangular matrix\n\n");
    printf("#### args: n=%" PRId64 ", lda=%" PRId64 ", incX=%" PRId64 ", order=%c, uplo=%c, trans=%c, diag=%c\n",
           (int64_t)N, (int64_t)lda, (int64_t)incX, order_to_char(order), uplo_to_char(uplo), transpose_to_char(trans),
           diag_to_char(diag));

    // allocate memory
    A = (float *)malloc(lda * N * sizeof(float));
    X = (float *)malloc(len_x * sizeof(float));

    // fill data
    fill_smatrix(A, N, N, lda, order, UPLO_TO_FILL_MODE(uplo), diag);
    fill_svector(X, N, incX);

    // print input data
    print_smatrix(A, N, N, lda, "A", order);
    print_svector(X, N, incX, "X");

    // call cblas_strsv
    cblas_strsv(order, uplo, trans, diag, N, A, lda, X, incX);

    // print result
    printf("\nAfter strsv operation:\n");
    print_svector(X, N, incX, "X");

    // release memory
    free(A);
    free(X);
    return 0;
}
